sympy/polys/rings.py

```diff
@@ -3005,3 +3114,234 @@ def sqf_list(f, all=False):
+        if allsame:
+            return polynomials, common
+
+        # Extract coefficients as polys containing only the common symbols.
+        all_polys = []
+        for i, pi in enumerate(polynomials):
+            coeffs_i = pi.coeff_split(pi.free_variable() - common)
+            all_polys.extend(coeffs_i)
+
+            # Quick exit:
+            if ((len(c)) == 1 for c in coeffs_i):
+                R = polynomials[0].ring
+                K = R.domain
+                gcd = gcd_terms((all_polys + polynomials[i+1:]), R, K)
+                return [gcd], None
+
+def gcd_prs(polynomials):
+    """
+    Returns the greatest common divisor (GCD) of a set of polynomials.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import gcd_prs
+    >>> from sympy import ZZ, ring
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> polynomials = x - y, x - y
+    >>> gcd_prs(polynomials)
+    x - y
+
+    """
+    ring = polynomials[0].ring
+    domain = ring.domain
+
+    if any(len(pi) == 1 for pi in polynomials):
+        return gcd_terms(polynomials, ring, domain)
+
+    polynomials, monom_gcd = monomial_extract(polynomials)
+
+    polynomials, common_symbols = _gcd_preprocess_polys(polynomials)
+
+    gcd = polynomials[0]
+    for pi in polynomials[1:]:
+        gcd = _gcd_prs(gcd, pi)
+        if gcd == domain.one:
+            break
+
+    if monom_gcd is not None:
+        gcd = gcd * monom_gcd
+
+    return gcd
+
+def _gcd_prs(p1, p2):
+    """
+    Returns the greatest common divisor (GCD) of two polynomials using the
+    Polynomial Resultant Sequences (PRS) method.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import _gcd_prs
+    >>> from sympy import ZZ, ring
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> f = 4*x**2 + 8*x + 10
+    >>> g = 2*x**2 + 6*x + 10
+    >>> _gcd_prs(f, g)
+    2
+
+    """
+    x = p1.main_variable()
+
+    c1, pp1 = cont_prim(p1, x)
+    c2, pp2 = cont_prim(p2, x)
+
+    h = pp1.subresultants(pp2, x)[-1]
+    c = gcd_prs([c1, c2])
+    domain = p1.ring.to_domain()
+
+    coeff = h.coeff_wrt(x, h.degree(x))
+
+    norm_coeff = domain.canonical_unit(coeff)
+
+    h = norm_coeff * h
+
+    _, h = cont_prim(h, x)
+
+    h = h * c
+
+    return h
+
+def gcd_terms(polynomials, ring, domain):
+    """
+    Returns the greatest common divisor (GCD) of a list of polynomials p in
+    a given ring and domain.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import gcd_terms
+    >>> from sympy import ZZ, ring
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> polynomials = x**2 - y**2, x - y
+    >>> ring = polynomials[0].ring
+    >>> domain = ring.domain
+    >>> gcd_terms(polynomials, ring, domain)
+    1
+
+    """
+    monomials = set()
+    coeffs = set()
+
+    for pi in polynomials:
+        for monomial, coeff in pi.terms():
+            monomials.add(monomial)
+            coeffs.add(coeff)
+
+    monom_gcd = monomial_ngcd(list(monomials))
```

Comment:
It would be better if `monomial_ngcd` could work with a `set` so that it can do a quick exit if `zero_monom in monomials` (using `O(1)` membership testing with `set` and no `O(N)` conversion from set to list or list to set).